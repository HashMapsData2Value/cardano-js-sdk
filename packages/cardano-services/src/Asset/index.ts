export * from './AssetBuilder';
export * from './AssetHttpService';
export * from './CardanoTokenRegistry';
export * from './DbSyncAssetProvider';
export * from './DbSyncNftMetadataService';
export * from './types';
