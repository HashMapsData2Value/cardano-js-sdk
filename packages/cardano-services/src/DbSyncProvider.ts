import { CardanoNode, HealthCheckResponse, Provider } from '@cardano-sdk/core';
import { Pool } from 'pg';

const HEALTH_CHECK_QUERY = 'SELECT 1';
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type AnyArgs = any[];
export const DbSyncProvider = <
  T extends (abstract new (...args: AnyArgs) => {}) | (new (...args: AnyArgs) => {}) = { new (): {} }
>(
  BaseClass?: T
) => {
  abstract class Mixin extends (BaseClass || Object) implements Provider {
    public db: Pool;
    public cardanoNode: CardanoNode;

    constructor(...args: AnyArgs) {
      const [db, cardanoNode, ...baseArgs] = [...args];

      super(...baseArgs);

      this.db = db;
      this.cardanoNode = cardanoNode;
    }

    public async healthCheck(): Promise<HealthCheckResponse> {
      // TODO: query block table, get last block minus 5
      const dbHealthCheck = { ok: !!(await this.db.query(HEALTH_CHECK_QUERY)).rowCount };
      const cardanoNodeHealthCheck = await this.cardanoNode.healthCheck();
      return {
        localNode: cardanoNodeHealthCheck.localNode ?? undefined,
        ok: dbHealthCheck.ok && cardanoNodeHealthCheck.ok
      };
    }
  }

  type BaseArgs = T extends abstract new (...baseArgs: infer A) => {}
    ? A
    : T extends new (...baseArgs: infer A) => {}
    ? A
    : never;
  type BaseInstance = T extends abstract new (...baseArgs: AnyArgs) => infer I
    ? I
    : T extends new (...baseArgs: AnyArgs) => infer I
    ? I
    : never;
  type ReturnedType = BaseInstance & {
    db: Pool;
    cardanoNode: CardanoNode;
    healthCheck: () => Promise<HealthCheckResponse>;
  };

  return Mixin as unknown as (T extends new (...baseArgs: AnyArgs) => {}
    ? new (db: Pool, cardanoNode: CardanoNode, ...args: BaseArgs) => ReturnedType
    : abstract new (db: Pool, cardanoNode: CardanoNode, ...args: BaseArgs) => ReturnedType) & {
    prototype: { healthCheck: () => Promise<HealthCheckResponse> };
  };
};
