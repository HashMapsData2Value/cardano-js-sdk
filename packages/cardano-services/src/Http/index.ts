export * from './HttpServer';
export * from './HttpService';
