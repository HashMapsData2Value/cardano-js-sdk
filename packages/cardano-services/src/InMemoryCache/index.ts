export * from './InMemoryCache';
export * from './defaults';
