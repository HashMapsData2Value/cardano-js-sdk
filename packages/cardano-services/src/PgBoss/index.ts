import { POOL_DELIST_SCHEDULE, PgBossQueue, PgBossSchedule, WorkerHandlerFactory } from './types';
import { STAKE_POOL_METADATA_QUEUE, STAKE_POOL_METRICS_UPDATE } from '@cardano-sdk/projection-typeorm';
import { stakePoolBatchDelistHandlerFactory } from './stakePoolBatchDelistHandler';
import { stakePoolMetadataHandlerFactory } from './stakePoolMetadataHandler';
import { stakePoolMetricsHandlerFactory } from './stakePoolMetricsHandler';

export * from './types';
export * from './util';

/** Defines the handler for each pg-boss queue */
export const queueHandlers: Record<PgBossQueue | PgBossSchedule, WorkerHandlerFactory> = {
  [POOL_DELIST_SCHEDULE]: stakePoolBatchDelistHandlerFactory,
  [STAKE_POOL_METADATA_QUEUE]: stakePoolMetadataHandlerFactory,
  [STAKE_POOL_METRICS_UPDATE]: stakePoolMetricsHandlerFactory
};
