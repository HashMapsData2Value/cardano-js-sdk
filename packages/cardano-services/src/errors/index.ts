export * from './InvalidLoggerLevel';
export * from './InvalidModuleState';
