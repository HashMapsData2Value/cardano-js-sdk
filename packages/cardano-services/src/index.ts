export * from './Http';
export * from './Program';
export * from './RunnableModule';
export * from './StakePoolSearch';
export * from './TxSubmit';
export * from './TxWorker';
export * from './Utxo';
export * as errors from './errors';
