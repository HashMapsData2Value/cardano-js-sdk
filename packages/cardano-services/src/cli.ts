#!/usr/bin/env node
/* eslint-disable import/imports-first */
require('../scripts/patchRequire');
import {
  API_URL_DEFAULT,
  OGMIOS_URL_DEFAULT,
  ProgramOptionDescriptions,
  ServiceNames,
  loadHttpServer
} from './Program';
import { Command } from 'commander';
import { InvalidLoggerLevel } from './errors';
import { LogLevel } from 'bunyan';
import { URL } from 'url';
import { loggerMethodNames } from './util';
import onDeath from 'death';
const clear = require('clear');
const packageJson = require('../package.json');

clear();
// eslint-disable-next-line no-console
console.log('Cardano Services CLI');

const program = new Command('cardano-services');

program.version(packageJson.version);

program
  .command('start-server')
  .description('Start the HTTP server')
  .argument('<serviceNames...>', `List of services to attach: ${Object.values(ServiceNames).toString()}`)
  .option('--api-url <apiUrl>', ProgramOptionDescriptions.ApiUrl, (url) => new URL(url), new URL(API_URL_DEFAULT))
  .option('--enable-metrics <metricsEnabled>', ProgramOptionDescriptions.MetricsEnabled, false)
  .option('--db-connection-string <dbConnectionString>', ProgramOptionDescriptions.DbConnection, (url) =>
    new URL(url).toString()
  )
  .option(
    '--ogmios-url <ogmiosUrl>',
    ProgramOptionDescriptions.OgmiosUrl,
    (url) => new URL(url),
    new URL(OGMIOS_URL_DEFAULT)
  )
  .option(
    '--logger-min-severity <level>',
    ProgramOptionDescriptions.LoggerMinSeverity,
    (level) => {
      if (!loggerMethodNames.includes(level)) {
        throw new InvalidLoggerLevel(level);
      }
      return level;
    },
    'info'
  )
  .action(
    async (
      serviceNames: ServiceNames[],
      {
        apiUrl,
        dbConnectionString,
        loggerMinSeverity,
        metricsEnabled,
        ogmiosUrl
      }: {
        apiUrl: URL;
        dbConnectionString?: string;
        loggerMinSeverity: LogLevel;
        metricsEnabled: boolean;
        ogmiosUrl?: URL;
      }
    ) => {
      const server = await loadHttpServer({
        apiUrl: apiUrl || API_URL_DEFAULT,
        options: {
          dbConnectionString,
          loggerMinSeverity,
          metricsEnabled,
          ogmiosUrl
        },
        serviceNames
      });
      await server.initialize();
      await server.start();
      onDeath(async () => {
        await server.shutdown();
        process.exit(1);
      });
    }
  );

if (process.argv.slice(2).length === 0) {
  program.outputHelp();
  process.exit(1);
} else {
  program.parseAsync(process.argv).catch((error) => {
    // eslint-disable-next-line no-console
    console.error(error);
    process.exit(0);
  });
}
