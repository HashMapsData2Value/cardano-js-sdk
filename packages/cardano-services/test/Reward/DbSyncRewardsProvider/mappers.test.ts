import { rewardsToCore } from '../../../src';

describe('DbSyncRewardProvider mappers', () => {
  const rewardsModels = [
    {
      address: 'stake_test1uzxvhl83q8ujv2yvpy6n2krvpdlqqx28h7e9vsk6re43h3c3kufy6',
      epoch: 10,
      quantity: '10000000'
    }
  ];
  it('rewardsToCore returns core types', () => {
    expect(rewardsToCore(rewardsModels)).toMatchSnapshot();
  });
  it('rewardsToCore empty rewards returns core types', () => {
    expect(rewardsToCore([])).toMatchSnapshot();
  });
  it('rewardsToCore returns core types indexed by address', () => {
    const rewards = [
      {
        address: 'stake_test1uzxvhl83q8ujv2yvpy6n2krvpdlqqx28h7e9vsk6re43h3c3kufy6',
        epoch: 11,
        quantity: '20000000'
      },
      {
        address: 'stake_test1uqfu74w3wh4gfzu8m6e7j987h4lq9r3t7ef5gaw497uu85qsqfy27',
        epoch: 15,
        quantity: '30000000'
      }
    ];
    expect(rewardsToCore([...rewardsModels, ...rewards])).toMatchSnapshot();
  });
});
