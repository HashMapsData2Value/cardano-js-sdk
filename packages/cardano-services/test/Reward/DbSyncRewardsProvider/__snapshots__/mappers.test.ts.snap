// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DbSyncRewardProvider mappers rewardsToCore empty rewards returns core types 1`] = `Map {}`;

exports[`DbSyncRewardProvider mappers rewardsToCore returns core types 1`] = `
Map {
  "stake_test1uzxvhl83q8ujv2yvpy6n2krvpdlqqx28h7e9vsk6re43h3c3kufy6" => Array [
    Object {
      "epoch": 10,
      "rewards": 10000000n,
    },
  ],
}
`;

exports[`DbSyncRewardProvider mappers rewardsToCore returns core types indexed by address 1`] = `
Map {
  "stake_test1uzxvhl83q8ujv2yvpy6n2krvpdlqqx28h7e9vsk6re43h3c3kufy6" => Array [
    Object {
      "epoch": 10,
      "rewards": 10000000n,
    },
    Object {
      "epoch": 11,
      "rewards": 20000000n,
    },
  ],
  "stake_test1uqfu74w3wh4gfzu8m6e7j987h4lq9r3t7ef5gaw497uu85qsqfy27" => Array [
    Object {
      "epoch": 15,
      "rewards": 30000000n,
    },
  ],
}
`;
